Code/GraphMol/FileParsers/GeneralFileReader.h

```diff
@@ -0,0 +1,213 @@
+#ifndef GENERAL_FILE_READER_H
+#define GENERAL_FILE_READER_H
+#include <string>
+#include <iostream>
+#include <vector>
+#include <boost/filesystem.hpp>
+#include "MolSupplier.h"
+#include <RDGeneral/BadFileException.h>
+#include <RDStreams/streams.h>
+
+namespace fs = boost::filesystem;
+
+namespace RDKit{
+	struct SupplierOption{
+		bool takeOwnership = true;
+		bool sanitize = true;
+		bool removeHs = true;
+		bool strictParsing = true;
+
+		std::string delimiter = "\t";
+		int smilesColumn = 0;
+		int nameColumn = 1;
+		bool titleLine = true;
+
+		std::string nameRecord = "";
+		int confId2D = -1;
+		int confId3D = 0;
+	};
+
+  class GeneralFileReader{
+	  public:
+			std::string fname;
+		  std::string file_format, compression_format;
+
+		struct SupplierOption opt;	// options for the Mol Supplier 
+		public:
+			GeneralFileReader(std::string fileName);
+		  GeneralFileReader(std::string fileName, struct SupplierOption options);
+
+		  /*
+		    Function to check the validity of the file and compression format
+	    */
+		  bool valid();
+
+		  /*
+		    Function to set the file and compression format
+		  */
+		  void setFormat();
+
+		  /*
+		    Get MolSupplier Object based on the file and compression format
+	    */
+		  RDKit::MolSupplier* getSupplier();
+  };
+
+	GeneralFileReader::GeneralFileReader(std::string fileName){
+		fname = fileName;
+		file_format = "";
+		compression_format = "";
+	}
+
+	GeneralFileReader::GeneralFileReader(std::string fileName, struct SupplierOption options){
+		fname = fileName;
+		file_format = "";
+		compression_format = "";
+		opt = options;
+	}
+
+	bool GeneralFileReader::valid(){
+
+		std::vector<std::string > file_formats{ "sdf", "mae", "smi", "csv", "txt", "tsv", "tdt"};
+		std::vector<std::string > compression_formats{ "gz" };
+
+    // Case 1: Unconventional format types
+		if(file_format.compare("maegz") == 0){
+			file_format = "mae";
+			compression_format = "gz";
+			return true;
+		}
+
+	
+		// Case 2: Either the filename has a file format or compression format or both
+		bool flag_fileFormat = std::find(file_formats.begin(), file_formats.end(), file_format) != file_formats.end();
+		
+		if (compression_format.compare("") != 0){
```

Comment:
this test is better as:
```suggestion
		if (!compression_format.empty()){
```
This comment applies to all the other uses of `.compare("")` in this review